### PR TITLE
Add Non-deterministic key derivation mode

### PR DESCRIPTION
**Pull request by sree-revoori1**
This commit introduces non-deterministic key derivation mode. See more details about the specification here -
https://github.com/TrustedComputingGroup/Server-Internal/blob/main/dpe-irot-profile/main.md#non-deterministic-mode-key-derivation.

### CODE REVIEW

File: dpe/src/commands/certify_key.rs [hunk 1]
```diff
@@ -19,21 +19,48 @@ pub struct CertifyKeyCmd {
...
+        if context.locality != locality {
             return Err(DpeErrorCode::InvalidHandle);
         }
 
-        // Derive CDI and public key
-        let cdi = dpe.derive_cdi(idx)?;
-        let pub_key = C::derive_ecdsa_pub(DPE_PROFILE.alg_len(), &cdi, &self.label, b"ECC")
-            .map_err(|_| DpeErrorCode::InternalError)?;
+        let algs = DPE_PROFILE.alg_len();
+        let pub_key = if context.cached_priv_key.is_some() {
```
Comment (zhalvorsen) [inline dpe/src/commands/certify_key.rs:49]: Should we only take the cached key if this request is also non-deterministic? In other words

```Rust
if self.uses_nd_derivation && context.cached_priv_key.is_some()
```

File: dpe/src/commands/sign.rs [hunk 2]
```diff
@@ -245,4 +278,92 @@ mod tests {
...
+                tci_type: i as u32,
+                target_locality: 0,
+            }
+            .execute(&mut dpe, TEST_LOCALITIES[0])
+            .unwrap();
+        }
+
+        let sig = {
+            // Check that private key is not cached and hence must be generated by sign
+            let idx = dpe
```
Comment (zhalvorsen) [inline dpe/src/commands/sign.rs:308]: fyi: Can pull this `idx` assignment out one layer so it can be used later.

File: dpe/src/support.rs [hunk 3]
```diff
@@ -149,7 +175,7 @@ pub mod test {
         .get_flags();
         assert_eq!(
             flags,
-            (1 << 31) | (1 << 29) | (1 << 27) | (1 << 25) | (1 << 23)
+            (1 << 31) | (1 << 29) | (1 << 27) | (1 << 25) | (1 << 21)
```
Comment (zhalvorsen) [inline dpe/src/support.rs:178]: fyi: this check was intended to check all of the odd-numbered bit flags.

File: dpe/src/support.rs [hunk 4]
```diff
@@ -159,7 +185,7 @@ pub mod test {
             ..Support::default()
         }
         .get_flags();
-        assert_eq!(flags, (1 << 30) | (1 << 28) | (1 << 26) | (1 << 24));
+        assert_eq!(flags, (1 << 30) | (1 << 28) | (1 << 26) | (1 << 22));
```
Comment (zhalvorsen) [inline dpe/src/support.rs:188]: fyi: this check was intended to check all of the even-numbered bit flags.

File: crypto/src/openssl.rs [hunk 5]
```diff
@@ -51,37 +52,21 @@ impl OpensslCrypto {
...
     }
 
-    pub fn derive_ecdsa_key(
+    fn ec_key_from_priv_key(
         algs: AlgLen,
-        cdi: &OpensslCdi,
-        label: &[u8],
-        info: &[u8],
-    ) -> EcKey<Private> {
+        priv_key: EcdsaPriv,
```
[the remark this answers is not in the document]
Comment (zhalvorsen) [reply crypto/src/openssl.rs]: As mentioned offline we should be able to with something similar to this in certify_key and sign:

```Rust
        let priv_key = if self.uses_nd_derivation() {
            dpe.contexts[idx].cached_priv_key.take().unwrap_or({
                let cdi = dpe.derive_cdi(idx, true)?;
                C::derive_ecdsa_key(algs, &cdi, &self.label, b"ECC")
            })
        } else {
            let cdi = dpe.derive_cdi(idx, false)?;
            C::derive_ecdsa_key(algs, &cdi, &self.label, b"ECC")
        };
        let pub_key = C::derive_ecdsa_pub(DPE_PROFILE.alg_len(), &priv_key)
            .map_err(|_| DpeErrorCode::InternalError)?;
        // cache private key
        if self.uses_nd_derivation() {
            let context = &mut dpe.contexts[idx];
            context.cached_priv_key.replace(priv_key);
        }
```

File: dpe/src/commands/certify_key.rs [hunk 6]
```diff
@@ -19,21 +19,50 @@ pub struct CertifyKeyCmd {
...
+        if context.locality != locality {
             return Err(DpeErrorCode::InvalidHandle);
         }
 
-        // Derive CDI and public key
-        let cdi = dpe.derive_cdi(idx)?;
-        let pub_key = C::derive_ecdsa_pub(DPE_PROFILE.alg_len(), &cdi, &self.label, b"ECC")
+        let algs = DPE_PROFILE.alg_len();
+        let priv_key = if self.uses_nd_derivation() {
+            dpe.contexts[idx].cached_priv_key.take().unwrap_or({
```
Comment (ia0) [inline dpe/src/commands/certify_key.rs:50]: I'm not sure of the expected behavior here, but we might want to use `unwrap_or_else()`. The difference being that if there's a cached private key, then we won't derive a new key. Right now we always generate a new key but simply discard it if there's a cached private key.

File: dpe/src/commands/sign.rs [hunk 7]
```diff
@@ -20,20 +20,25 @@ pub struct SignCmd {
...
 
     const fn _uses_symmetric(&self) -> bool {
         self.flags & Self::_IS_SYMMETRIC != 0
     }
 
     /// Uses non-deterministic derivation. If symmetric algorithms are used, this flag is ignored.
-    const fn _uses_nd_derivation(&self) -> bool {
-        !self._uses_symmetric() && self.flags & Self::_ND_DERIVATION != 0
+    const fn uses_nd_derivation(&self) -> bool {
+        !self._uses_symmetric() && self.flags & Self::ND_DERIVATION != 0
```
Comment (ia0) [inline dpe/src/commands/sign.rs:31]: nit: Now that this function is used we can also remove the underscore prefix from `uses_symmetric`. But could/should be a separate PR.

File: dpe/src/commands/sign.rs [hunk 8]
```diff
@@ -46,12 +51,26 @@ impl<C: Crypto> CommandExecution<C> for SignCmd {
...
 
-        let cdi = dpe.derive_cdi(idx)?;
         let digest = Digest::new(&self.digest, DPE_PROFILE.alg_len())
             .map_err(|_| DpeErrorCode::InternalError)?;
-        let EcdsaSig { r, s } =
-            C::ecdsa_sign_with_derived(DPE_PROFILE.alg_len(), &cdi, &self.label, b"ECC", &digest)
-                .map_err(|_| DpeErrorCode::InternalError)?;
+        let algs = DPE_PROFILE.alg_len();
+        let priv_key = if self.uses_nd_derivation() {
+            dpe.contexts[idx].cached_priv_key.take().unwrap_or({
```
Comment (ia0) [inline dpe/src/commands/sign.rs:58]: Same comment as previously regarding using `unwrap_or_else()`.

File: dpe/src/commands/sign.rs [hunk 9]
```diff
@@ -46,12 +51,26 @@ impl<C: Crypto> CommandExecution<C> for SignCmd {
...
+            let cdi = dpe.derive_cdi(idx, false)?;
+            C::derive_ecdsa_key(algs, &cdi, &self.label, b"ECC")
+        };
+
+        let EcdsaSig { r, s } = C::ecdsa_sign_with_derived(algs, &digest, &priv_key)
+            .map_err(|_| DpeErrorCode::InternalError)?;
+        // cache private key
+        if self.uses_nd_derivation() {
+            let context = &mut dpe.contexts[idx];
+            context.cached_priv_key.replace(priv_key);
```
Comment (ia0) [inline dpe/src/commands/sign.rs:72]: nit: Not sure introducing a variable helps here. I would expect the same usage as line 58.

```rust
dpe.contexts[idx].cached_priv_key.replace(priv_key);
```

**Review by ia0 — COMMENTED** [review]
LGTM at Rust-level (modulo the `unwrap_or_else()` thing)

File: crypto/src/lib.rs [hunk 10]
```diff
@@ -159,21 +162,17 @@ pub trait Crypto {
...
-    /// * `info` - Caller-supplied info string to use in asymmetric key derivation
     /// * `digest` - Digest of data to be signed.
+    /// * `priv_key` - Caller-supplied private key to use in public key derivation
     fn ecdsa_sign_with_derived(
         algs: AlgLen,
-        cdi: &Self::Cdi,
-        label: &[u8],
-        info: &[u8],
         digest: &Digest,
+        priv_key: &EcdsaPriv,
```
Comment (jhand2) [inline crypto/src/lib.rs:175]: So the reason this takes a cdi instead of a private key is that in some integrations they actual private key will be held in hardware and firmware will never see the private key.

In the Caliptra implementation of this trait, I think the Cdi will just be the index of a KeyVault slot.

So at least for now I think it's better to rederive the key from the CDI whenever it is needed.

File: crypto/src/openssl.rs [hunk 11]
```diff
@@ -116,37 +101,70 @@ impl Crypto for OpensslCrypto {
...
+                    &KdfArgument::KbMode(KdfKbMode::Counter),
+                    &KdfArgument::Mac(KdfMacType::Hmac(md)),
+                    &KdfArgument::KbInfo(measurement.bytes()),
+                    &KdfArgument::Salt(info),
+                    &KdfArgument::Key(&measurement.bytes()),
+                ];
+                perform_kdf(KdfType::KeyBased, &args, md.size())
+                    .map_err(|_| CryptoError::CryptoLibError)
+            }
+        }
```
Comment (jhand2) [inline crypto/src/openssl.rs:132]: nit: could maybe clean this up by doing something like this. I'm not 100% sure if this works though.

```rust
let mut args = vec![
    &KdfArgument::KbMode(KdfKbMode::Counter),
    &KdfArgument::Mac(KdfMacType::Hmac(md)),
    &KdfArgument::KbInfo(measurement.bytes()),
    &KdfArgument::Salt(info),
    &KdfArgument::Key(&measurement.bytes()),
];

if let Some(seed) = rand_seed {
    args.extend(&KdfArgument::KbSeed(seed));
}

perform_kdf(KdfType::KeyBased, args.as_slice(), md.size())
       .map_err(|_| CryptoError::CryptoLibError)
```

File: crypto/src/openssl.rs [hunk 12]
```diff
@@ -116,37 +101,70 @@ impl Crypto for OpensslCrypto {
...
+    }
+
+    fn derive_ecdsa_key(algs: AlgLen, cdi: &Self::Cdi, label: &[u8], info: &[u8]) -> EcdsaPriv {
+        let md = Self::get_digest(algs);
         let args = [
             &KdfArgument::KbMode(KdfKbMode::Counter),
             &KdfArgument::Mac(KdfMacType::Hmac(md)),
-            &KdfArgument::KbInfo(measurement.bytes()),
+            &KdfArgument::KbInfo(label),
```
Comment (jhand2) [inline crypto/src/openssl.rs:140]: I'll come back to this, once the first comment is addressed.

Comment (jhand2) [reply crypto/src/openssl.rs:140]: No, not applicable anymore

File: crypto/src/openssl.rs (hunk 11 above)
Comment (jhand2) [reply crypto/src/openssl.rs:132]: Ok, sounds good

**Review by jhand2 — APPROVED** [review]
